tuta-sdk/rust/src/crypto/entity_facade.rs

```diff
@@ -0,0 +1,304 @@
+                        if let ElementValue::Null = association {
+                            ElementValue::Null
+                        } else {
+                            let decrypted_aggregate = self.decrypt_and_map(aggregate_type_model, association.assert_dict(), session_key);
+                            match decrypted_aggregate {
+                                Ok(dec_aggregate) => {
+                                    let mut aggregate = dec_aggregate.clone();
+                                    self.extract_errors(association_name, &mut errors, &mut aggregate);
+                                    ElementValue::Dict(aggregate)
+                                }
+                                Err(_) => {
+                                    return Err(ApiCallError::InternalSdkError { error_message: format!("Failed to decrypt association {association_name}") });
+                                }
+                            }
+                        }
+                    }
+                    None => ElementValue::Null
+                };
+
+                Ok((decrypted_aggregate, errors))
+            }
+        } else {
+            let value = entity.get(association_name).unwrap_or(&ElementValue::Null);
+            Ok((value.clone(), errors))
+        };
+    }
+
+    fn extract_errors(&self, association_name: &String, errors: &mut HashMap<String, ElementValue>, dec_aggregate: &mut ParsedEntity) {
+        if dec_aggregate.get("errors").is_some() {
+            errors.insert(association_name.to_string(), ElementValue::Dict(dec_aggregate.get("errors").unwrap().assert_dict()));
+            dec_aggregate.remove("errors");
+        }
+    }
+
+    fn map_value(&self, entity: ParsedEntity, session_key: &AesKey, key: &String, model_value: &ModelValue) -> Result<(ElementValue, HashMap<String, ElementValue>, HashMap<String, ElementValue>), ApiCallError> {
+        let mut final_ivs: HashMap<String, ElementValue> = Default::default();
+        let mut errors: HashMap<String, ElementValue> = Default::default();
+
+        let value = entity.get(key.as_str()).unwrap_or_else(|| &ElementValue::Null);
+
+        if model_value.encrypted && model_value.is_final && value != &ElementValue::Null {
+            final_ivs.insert(key.to_string(), ElementValue::Bytes(self.extract_iv(value.assert_bytes().as_slice()).to_vec()));
+        }
+
+        if value == &ElementValue::Null {
+            if model_value.cardinality != Cardinality::ZeroOrOne {
+                return Err(ApiCallError::InternalSdkError { error_message: format!("Value {key} with cardinality ONE can't be null") });
+            }
+
+            return Ok((ElementValue::Null, final_ivs, errors));
+        } else if model_value.cardinality == Cardinality::One && value == &ElementValue::String(String::new()) {
+            return Ok((self.resolve_default_value(model_value.value_type.to_owned()), final_ivs, errors));
+        }
+
+        if model_value.encrypted {
```

Comment:
I think I would return the whole `if` with `else`, it's confusing to me that there's no `else` to be found